### PR TITLE
upgrade deseq2 and include tximport to use sailfish results directly 

### PR DESCRIPTION
**Pull request by bgruening**
This one is called DESeq2-TNG I should have named it RNA-seq TNG ;)
I'm very excited about this PR and I'm seeking for help in testing. A little bit of background: until now we need to map the reads and use htseq-count to count reads in a region to feed this into DESeq2.
On the other hand Sailfish is a "rapid Alignment-free Quantification of Isoform Abundance" tool ... but we can not use the output natively with DESeq2. This has changed now!
@mikelove has developed a nice tool called tximport that can read sailfish and salmon outputs directly and does the DESEq2 magic on top of it.

This PR updated our DESEq2 wrapper to make use of this new functionality. The workflow should be much simpler and much faster, as it is alignment free. I have created a new DESEq2 conda package to ease testing. So please activate the "bgruening" conda channel and run planemo on it if you want to provide feedback!

@mikelove and me would love to have as much feedback as possible until the new bioc release in two weeks and the new DESeq2 release. After this there is a freeze for the next 6 month ... more or less.

Let us know what you think we both are pretty excited about this :)
Cheers,
Michael and Bjoern

xref: https://github.com/galaxyproject/tools-iuc/pull/699

### CODE REVIEW

**Comment by mikelove** [comment]
hi Bjoern, I'm wondering, should we move deseq2.R outside of the DESeq2 repo? It's pretty unfortunate that we have to wait 6 months between Bioc releases to do bug fixing. We could put it in a separate github repo? Or in the Galaxy wrapper repo and I could maintain it there?

**Comment by mvdbeek** [comment]
@mikelove @bgruening I like the idea of having deseq2.R in the iuc repo.

**Comment by yhoogstrate** [comment]
Adding it to the iuc (iuc-tools) repo means that for every change and update @mikelove has to go to our review process. It's up to him to decide if that's a desired working procedure? If the R file is specific for galaxy I think it's more practical in here, but if @mikelove may have plans to include/integrate it into other projects I think it's handier to maintain it in another repo.

**Comment by mikelove** [comment]
hi, thanks for the comments. It is Galaxy-specific, but maybe it is best if updates are mediated by Galaxy experts rather than me. @bgruening what if I put this script in its own Github repo, which would be the "official DESeq2 for Galaxy script"?

**Comment by yhoogstrate** [comment]
Also for the reason @lparsons gives in #716 and that it's specific for galaxy I think it's easiest to include it in the iuc-tools wrapper here, but if you prefer a separate repo it's fine as well

**Comment by mikelove** [comment]
I don't have any preference. Whatever makes most sense for the Galaxy folks. I'm happy to help maintain the script wherever. I would vote for moving it out of DESeq2 release branch so that we can have continual bug fixes.

**Comment by bgruening** [comment]
@mikelove ok then let's move it back into this repo here. Maybe this is the easiest way for all of us. I'm really happy that you help us and want to maintain it here! Thanks a lot!

**Comment by mikelove** [comment]
@bgruening let me know when you pull it from 

https://github.com/Bioconductor-mirror/DESeq2/blob/master/inst/script/deseq2.R

and then I will remove it from the DESeq2 svn

**Comment by bgruening** [comment]
Sorry, travelling right now. I see a few steps missing for this wrapper and it would be nice if someone can help here. Not sure If I find time the next week.

* [ ] moving the script back into this repo
* [ ] creating tool_dependency.xml file as soon as a new BioC version comes out
* [ ] updating conda packages with new urls
* [ ] use metadata column names 
* [ ] improve help somehow ( @lparsons  )

**Comment by bgruening** [comment]
@mikelove done :)

**Comment by mikelove** [comment]
great. I've removed it from Bioc devel branch now to avoid any confusion/duplicate versions.

**Comment by bgruening** [comment]
@yhoogstrate the column headers are not working as it seems. Did I something wrong here?

**Comment by yhoogstrate** [comment]
Hey @bgruening, I am looking into this. At first glance I think you need to encapsulate the `<action ...>` in between `<actions><action ... /></actions>`, and I should obviously update the wiki.

**Comment by yhoogstrate** [comment]
We're also 1 column too short. Probably the second column, marked as Log2FC, which is always positive in my the test-data's test case.

screenshot from 2016-04-13 21 05 20

**Comment by mikelove** [comment]
@yhoogstrate column 2 is "base mean", the average of normalized counts across all samples

here's an example results table from DESeq2:

http://master.bioconductor.org/help/workflows/rnaseqGene/#building-the-results-table

**Comment by bgruening** [comment]
Awesome! Thanks for the review @yhoogstrate.
This works pretty well in our testing. For the record I just wait for the Bioc release and the dependencies. So we can merge this. I will work on conda deps anyone wants to create tool_depedencies.xml?

**Comment by yhoogstrate** [comment]
I have added a few small textual changes. Revert them if you don't like them. If I understand it correctly  https://github.com/galaxyproject/tools-iuc/pull/690 is doing more or less the same but is more outdated. Should we close that one and only focus on this PR?

File: tools/deseq2/deseq2.xml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<tool id="deseq2" name="DESeq2" version="2.11.38">
+<tool id="deseq2" name="DESeq2" version="2.1.11.38">
```
Comment (bgruening) [inline tools/deseq2/deseq2.xml]: Why is this needed?

File: tools/deseq2/deseq2.R [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+# otherwise, from sample_table, assume it is relative
+dir <- if (is.null(opt$factors)) {
+  "."
+} else {
+  ""
+}
+
+if (!useTXI) {
+  # construct the object from HTSeq files
+  dds <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,
```
Comment (yhoogstrate) [inline tools/deseq2/deseq2.R]: Is this the line causing #721 ?

**Comment by mikelove** [comment]
No, i think that issue is from within a DESeq2 function, which is actually just carried over from DESeq (1)

I don't know about any gene IDs with a space. What are these actually? It's a bad idea to have a space inside an ID

Maybe we can just throw an error within the Galaxy script if the gene IDs contain a space

**Comment by yhoogstrate** [comment]
I think @mikelove has a fair point. I would expect such problems to occur more often due to tabs being replaced by spaces or files that are in other ways corrupt instead of actual gene-ids. However, if there are gene IDs containing spaces we should consider fixing this.

**Comment by lparsons** [comment]
I agree that gene and ID should probably not contain a space, but in at least one case, I've run into that (UCSC galGal4 genome, refseq ids). Certainly, reading a tab delimited file shouldn't be a problem. If downstream things require no spaces, I would think simply replacing the spaces with underscore "_" would help to clean up the input and follow the [Robustness principle](https://en.wikipedia.org/wiki/Robustness_principle). Not sure exactly where the code is to fix, however.

**Comment by mikelove** [comment]
I guess the line of code is here:

https://github.com/Bioconductor-mirror/DESeq2/blob/master/R/AllClasses.R#L310

but this has been in DESeq since a long long long time and I've never heard of an issue:

https://github.com/Bioconductor-mirror/DESeq/blob/master/R/methods.R#L298-L299

I wouldn't want to tweak things without being totally sure that I'm not breaking anything.

Can you confirm which gene IDs contain a space? These Chicken refseq IDs:

https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=490299411_C1B0aHquTVHzqe9grMbOjRVkToxP&hgta_database=galGal4&hgta_histoTable=refGene&hgta_doValueHistogram=name

or these gene IDs?

https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=490299411_C1B0aHquTVHzqe9grMbOjRVkToxP&hgta_database=galGal4&hgta_histoTable=refGene&hgta_doValueHistogram=name2

**Comment by lparsons** [comment]
@mikelove The second link you sent ([galGal4 refGene name2](https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=490299411_C1B0aHquTVHzqe9grMbOjRVkToxP&hgta_database=galGal4&hgta_histoTable=refGene&hgta_doValueHistogram=name2)) contains the "id" `SET domain containing 5`.

I ended up with this when getting the GTF file from UCSC using Kent Tools `genePredToGtf galGal4 refGene galGal4_refGene.gtf`. I always use this method since the Table Browser doesn't generate correct GTF files.

**Comment by mikelove** [comment]
Wow! I think that's actually an error at UCSC? The official symbol is SETD5:

http://www.ncbi.nlm.nih.gov/gene/55209

The other values in the table are symbols.

Am I seeing it wrong?

**Comment by lparsons** [comment]
@mikelove I couldn't say for sure, but it looks pretty odd to me. I'm not sure that UCSC declares anywhere exactly what those fields contain, the authoritative source for that data, and the version/date they were obtained, so it's hard to say it's for sure what it should be. That is the "Official Full Name" according to that link.

**Comment by mikelove** [comment]
@lparsons My bad. I didn't specify organism. The "symbol" indeed has spaces and looks like a name, and the "full name" looks like the symbol:

http://www.ncbi.nlm.nih.gov/gene/416046

**Comment by mikelove** [comment]
I'm continuing down the rabbit hole. BirdBase has the symbol as would be expected, SETD5

http://birdbase.arizona.edu/birdbase/gene.jsp?db_key_value=457965

The source, Chicken Gene Nomenclature Consortium, is unfortunately down at the moment:

http://birdgenenames.org/cgnc/

**Comment by yhoogstrate** [comment]
@mikelove I think you crashed it because it can't deal with spaces in gene symbols :laughing:

**Comment by yhoogstrate** [comment]
@bgruening any idea why the test is failing with these errors?
```
Install conda target CondaTarget[bioconductor-deseq2,version=1.11.38]
/home/travis/conda/bin/conda create -y --name __package__bioconductor-deseq2@__version__1.11.38 bioconductor-deseq2=1.11.38
Fetching package metadata: ............
Solving package specifications: ...........................................................An unexpected error has occurred, please consider sending the
```
...
```
File "/home/travis/conda/lib/python2.7/site-packages/conda/resolve.py", line 151, in __init__
    raise ValueError(message + "invalid character(s).")
ValueError: Malformed version string '0.2-5': invalid character(s).
```
...
```
---------------------- >> begin tool stderr << -----------------------
/tmp/tmpTXe0qT/job_working_directory/000/8/tool_script.sh: line 9: R: command not found
/tmp/tmpTXe0qT/job_working_directory/000/8/tool_script.sh: line 9: Rscript: command not found
----------------------- >> end tool stderr << ------------------------
```

local installation and testing does work

**Comment by bgruening** [comment]
Seems that bioc and DESeq2 release has been postponed.

**Comment by bgruening** [comment]
DESeq2 was released with a new bioc release. I will create new conda packages as soon as possible and then we can get this in. Anyone interested in tool_dependencies.xml?
Thanks!

**Comment by mr-c** [comment]
@bgruening Can I see a sample of the JSON that Galaxy produces?

File: tools/deseq2/deseq2.xml [hunk 3]
```diff
@@ -1,8 +1,8 @@
-<tool id="deseq2" name="DESeq2" version="2.1.8.3">
+<tool id="deseq2" name="DESeq2" version="2.11.38">
     <description>Determines differentially expressed features from count tables</description>
     <requirements>
         <!-- odering is crucial, otherwise R will override the ENV variables from deseq2 -->
-        <requirement type="package" version="1.8.2">deseq2</requirement>
+        <requirement type="package" version="1.11.38">bioconductor-deseq2</requirement>
```
Comment (mvdbeek) [inline tools/deseq2/deseq2.xml]: what's going on with this version?
Is this to be released?
I can't seem to find it, instead i'm getting the bioconductor-deseq2-1.10.1-r3.2.2_0 through the versionless resolver.

Comment (bgruening) [inline tools/deseq2/deseq2.xml]: :( It's blocking on: https://github.com/ContinuumIO/anaconda-issues/issues/817

**Comment by mikelove** [comment]
anything I can do to help? Would be cool to have a Sailfish => DGE pipeline in Galaxy :)

**Comment by bgruening** [comment]
@galaxyproject/iuc I would be delighted to get a final review here!
We have also updated sailfish and salmon is nearly ready.

@mikelove not sure what do say. I feel ashamed that this took to long - as an excuse we have worked hard to solve the installation and distribution problem for all packages (not only R ones) - and we are now generating Docker containers for all packages as well, including tximport and deseq2.

From here on updating tools should be a blast and very easy, promised!

**Comment by mikelove** [comment]
@bgruening thank you for your hard work. This should be an exciting offering for Galaxy users. 

no worries, I could see that what was going on was general package solution-ing.

**Comment by mvdbeek** [comment]
I'll upload this to the TTS, for easier testing.

**Comment by bgruening** [comment]
@mvdbeek thanks a lot!

**Comment by mvdbeek** [comment]
I just uploaded it, I haven't gotten around to testing it yet, but I'm eager to get this in!

File: tools/deseq2/deseq2.xml [hunk 4]
```diff
@@ -83,9 +98,33 @@
...
+                <option value="count" selected="True">Count data (e.g. from htseq-count or feature-count)</option>
+                <option value="tximport">TPM values (e.g. from sailfish or salmon)</option>
+            </param>
+            <when value="tximport">
+                <conditional name="mapping_format">
+                    <param name="mapping_format_selector" type="select" label="Gene mapping format">
+                        <option value="gtf" selected="True">GTF</option>
+                        <option value="tabular">Transcript-ID and Gene-ID mapping file</option>
+                    </param>
+                    <when value="gtf">
```
Comment (bgruening) [inline tools/deseq2/deseq2.xml]: True. @yhoogstrate any idea? We could insert a  `no gene mapping` option in the conditional instead?

Comment (mvdbeek) [inline tools/deseq2/deseq2.xml]: My understanding is that one of the two is required, according to line 92 in tools/deseq2/deseq2.R. So you could just remove the optional="true", right?

Comment (bgruening) [inline tools/deseq2/deseq2.xml]: This would be possible, but I guess @yhoogstrate was asking if this should maybe entirely optional? No? I will remove optional for now.

**Comment by mvdbeek** [comment]
Very nice, I just tried with a multi-factor experiment with count files from featurecounts,
all seems to be working well!
:+1:

**Comment by bgruening** [comment]
Cool, thanks for the review @mvdbeek!

@mikelove and all others thanks a lot. Sailfish is also updated to the latest version and salmon is nearly ready.

File: tools/deseq2/deseq2.xml (hunk 4 above)
Comment (yhoogstrate) [inline tools/deseq2/deseq2.xml]: @bgruening you got it, thanks for fixing it too :+1: , thanks for reviewing @mvdbeek :+1:
